plugin/evm/validators_transitioner_test.go

```diff
@@ -0,0 +1,161 @@
+package evm
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/flare-foundation/flare/ids"
+)
+
+type testValidatorRetriever struct {
+	ByEpochFunc func(epoch uint64) (map[ids.ShortID]uint64, error)
+}
+
+func (t *testValidatorRetriever) ByEpoch(epoch uint64) (map[ids.ShortID]uint64, error) {
+	return t.ByEpochFunc(epoch)
+}
+
+func TestValidatorsTransitioner_ByEpoch(t *testing.T) {
+
+	t.Run("nominal case 1: includes check for recursion and cache usage", func(t *testing.T) {
+		t.Parallel()
+
+		calls := 0
+		testEpoch := uint64(10)
+		providers := fakeProviders(testEpoch)
+
+		mock := &testValidatorRetriever{
+			ByEpochFunc: func(epoch uint64) (map[ids.ShortID]uint64, error) {
+				calls++
+				return providers, nil
+			},
+		}
+
+		validatorsTransitioner := NewValidatorsTransitioner(nil, mock)
+
+		got, err := validatorsTransitioner.ByEpoch(testEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+		assert.Equal(t, 10, calls)
+
+		// calls is expected to increase by 1 only because the cache would be used
+		got, err = validatorsTransitioner.ByEpoch(testEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+		assert.Equal(t, 11, calls)
+
+		mock.ByEpochFunc = func(epoch uint64) (map[ids.ShortID]uint64, error) {
+			calls++
+			return providers, nil
+		}
+
+		got, err = validatorsTransitioner.ByEpoch(testEpoch + 1)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+		assert.Equal(t, 13, calls)
+	})
+
+	t.Run("nominal case 2: non-nil default validators to reach the end of the function", func(t *testing.T) {
+		t.Parallel()
+
+		testEpoch := uint64(10)
+		providers := fakeProviders(testEpoch)
+		// here we give non-nil validators so that the later part of the ByEpoch() function can be reached by control and therefore tested
+		validators := fakeValidators(testEpoch)
+
+		mock := &testValidatorRetriever{
+			ByEpochFunc: func(epoch uint64) (map[ids.ShortID]uint64, error) {
+				return providers, nil
+			},
+		}
+
+		validatorsTransitioner := NewValidatorsTransitioner(validators, mock)
+
+		got, err := validatorsTransitioner.ByEpoch(testEpoch)
+		require.NoError(t, err)
+		assert.Equal(t, providers, got)
+
+		var totalWeight uint64
+		for _, u := range got {
+			totalWeight += u
+		}
+		assert.Equal(t, uint64(45), totalWeight)
```

Comment:
```suggestion
		testEpoch := uint64(10)
		wantWeight := uint64(45)
		providers := fakeProviders(testEpoch)
		// here we give non-nil validators so that the later part of the ByEpoch() function can be reached by control and therefore tested
		validators := fakeValidators(testEpoch)

		mock := &testValidatorRetriever{
			ByEpochFunc: func(epoch uint64) (map[ids.ShortID]uint64, error) {
				return providers, nil
			},
		}

		validatorsTransitioner := NewValidatorsTransitioner(validators, mock)

		got, err := validatorsTransitioner.ByEpoch(testEpoch)
		require.NoError(t, err)
		assert.Equal(t, providers, got)

		var totalWeight uint64
		for _, u := range got {
			totalWeight += u
		}
		assert.Equal(t, wantWeight , totalWeight)
```

Small nitpick but it's usually better to define "magic" expected output variables at the top of the test so that the reader has a clear idea of the ins and outs of the test while they read it IMO (but it's debatable, that's just my personal preference 😄)